packages/integration_test/android/src/main/java/dev/flutter/plugins/integration_test/FlutterDeviceScreenshot.java

```diff
@@ -0,0 +1,174 @@
+   *
+   * @return true if the app is running with instrumentation.
+   */
+  static boolean hasInstrumentation() {
+    try {
+      return InstrumentationRegistry.getInstrumentation() != null;
+    } catch (IllegalStateException exception) {
+      return false;
+    }
+  }
+
+  /**
+   * Captures a screenshot by drawing the view to a Canvas.
+   *
+   * <p>It also converts {@link FlutterView} to an image view, since {@link FlutterSurfaceView}
+   * pixels are opaque when rendering the view to a canvas.
+   *
+   * @param activity This is {@link FlutterActivity}.
+   * @param methodChannel The method channel to call into Dart.
+   * @param result The result of the method call that came from Dart.
+   */
+  static void captureView(
+      @NonNull Activity activity, @NonNull MethodChannel methodChannel, @NonNull Result result) {
+    final FlutterView flutterView = getFlutterView(activity);
+    flutterView.convertToImageView();
+    methodChannel.invokeMethod("scheduleFrame", null);
+
+    final HandlerThread screenshotBackgroundThread = new HandlerThread("screenshot");
+    screenshotBackgroundThread.start();
+
+    final Handler backgroundHandler = new Handler(screenshotBackgroundThread.getLooper());
+    final Handler mainHandler = new Handler(Looper.getMainLooper());
+
+    takeScreenshot(backgroundHandler, mainHandler, flutterView, result);
+  }
+
+  /**
+   * Captures a screenshot using ui automation.
+   *
+   * @return byte array containing the screenshot.
+   */
+  static byte[] captureWithUiAutomation() throws IOException {
+    final Instrumentation instrumentation = InstrumentationRegistry.getInstrumentation();
+    final Bitmap originalBitmap = instrumentation.getUiAutomation().takeScreenshot();
+    final ByteArrayOutputStream output = new ByteArrayOutputStream();
+
+    originalBitmap.compress(Bitmap.CompressFormat.PNG, /* irrelevant for PNG */ 100, output);
+    return output.toByteArray();
+  }
+
+  private static FlutterView getFlutterView(Activity activity) {
+    final ViewGroup root = (ViewGroup) activity.findViewById(android.R.id.content);
+    return (FlutterView) (((ViewGroup) root.getChildAt(0)).getChildAt(0));
+  }
+
+  private static void waitForAndroidFrame(Runnable r) {
+    Choreographer.getInstance()
+        .postFrameCallback(
+            new Choreographer.FrameCallback() {
+              @Override
+              public void doFrame(long frameTimeNanos) {
+                r.run();
+              }
+            });
+  }
+
+  private static void takeScreenshot(
+      @NonNull Handler backgroundHandler,
+      @NonNull Handler mainHandler,
+      @NonNull FlutterView view,
+      @NonNull Result result) {
+    final boolean acquired = view.acquireLatestImageViewFrame();
+    // The next frame may already have already been comitted.
+    // The next frame is guaranteed to have the Flutter image.
+    waitForAndroidFrame(
+        () -> {
+          waitForAndroidFrame(
```

Comment:
This seems to mean that we'll pump at least two frames after calling this method, right?

How will this be stable for screenshot tests?

Reply:
For image views: Thsi should work because we'll know we've had at least one frame drawn by flutter.

For instrumentation tests: This should work because the Dart part of the test should have gotten us to a stable non-animating point in the application, and now should be waiting before tapping or doing anything else.

@blasten - will the new method block if awaited until the screenshot is finished?

Reply:
Removed support for instrumentation tests until https://github.com/flutter/flutter/issues/56591 is addressed.

It rasterizes in a separate background thread, so it doesn't block the main thread.

Reply:
In that case, do we need to fail if `!acquired`?

Reply:
This still seems to be an issue - I'm not sure if I'm just missing somethign here, but it seems like if `acquired` is false we'll fail to produce a screenshot.

Reply:
It’s recursive, so it will keep waiting.

Reply:
Ahh I get it now. Ok.

Will the test just timeout if they failed to convert to imageview first?

Reply:
Right